src/Relationships/OneHasMany.php

```diff
@@ -15,31 +15,40 @@ class OneHasMany extends HasMany
 {
 	public function getEntitiesForPersistence()
 	{
-		$entities = [];
-		foreach ($this->toAdd as $add) {
-			$entities[] = $add;
-		}
-		foreach ($this->toRemove as $remove) {
-			if ($remove->isPersisted()) {
-				$entities[] = $remove;
-			}
-		}
 		if ($this->collection !== null || $this->wasLoaded) {
-			foreach ($this->getIterator() as $entity) {
-				$entities[] = $entity;
+			$entities = iterator_to_array($this->getIterator());
+
+		} else {
+			$entities = $this->added + $this->toAdd;
+
+			foreach ($this->toRemove as $hash => $remove) {
+				if ($remove->isPersisted()) {
+					$entities[$hash] = $remove;
+				} else {
+					unset($entities[$hash]);
+				}
 			}
 		}
+
 		return $entities;
 	}
 
 
 	public function doPersist()
 	{
+		if (!$this->isModified) {
+			return;
+		}
+
+		if ($this->collection !== null) {
+			$this->collection = $this->applyDefaultOrder($this->collection); // required when ordered by id
```

Comment:
@JanTvrdik do you have idea why this is neede and why you have put it here? I see some tests fail if I removed so, something has changed in this pr to require this.